src/consumer.js

```diff
@@ -0,0 +1,229 @@
+const amqplib = require('amqplib');
+const assert = require('assert');
+const slugid = require('slugid');
+
+/**
+ * A PulseQueue declares a queue and listens for messages on that
+ * queue, invoking a callback for each message.
+ */
+class PulseQueue {
+  constructor({client, bindings, handleMessage, ...options}) {
+    this.client = client;
+    this.bindings = bindings;
+    this.handleMessage = handleMessage;
+    this.options = {prefetch: 5, ...options};
+
+    this._handleConnection = this._handleConnection.bind(this);
+
+    // false once stop() has been called
+    this.running = true;
+
+    // the current channel and consumerTag, if any
+    this.channel = null;
+    this.consumerTag = null;
+
+    // number of messages being processed right now, and a function to call
+    // when that number goes to zero
+    this.processingMessages = 0;
+    this.idleCallback = null;
+
+    // set up to call _handleConnection on new connections and, if there
+    // is already an active connection, on that as well.
+    this.client.on('connected', this._handleConnection);
+    if (this.client.activeConnection) {
+      this._handleConnection(this.client.activeConnection);
+    }
+  }
+
+  /**
+   * Stop listening.  After this call, no further messages will be consumed
+   * from the queue.  The queue and any bindings will remain configured on the
+   * server.  This method will return after any pending consumers have
+   * completed (ack or nack).
+   */
+  async stop() {
+    if (!this.running) {
+      return;
+    }
+    this.running = false;
+
+    return this._shutdown();
+  }
+
+  /**
+   * Shut down this listener and channel, without setting this.running to false
+   */
+  async _shutdown() {
+    const {channel, consumerTag} = this;
+
+    if (channel && consumerTag) {
+      this.consumerTag = null;
+      await channel.cancel(consumerTag);
+    }
+
+    // if messages are being processed, arrange to continue when they
+    // are all handled
+    if (this.processingMessages > 0) {
+      await new Promise(resolved => {
+        this.idleCallback = resolved;
+      });
+    }
+
+    if (channel) {
+      this.channel = null;
+      await channel.close();
+    }
+  }
+
+  /**
+   * Handle a new connection to the pulse server, re-declaring everything
+   * and setting up the consumer.
+   */
+  async _handleConnection(conn) {
+    try {
+      if (!this.running) {
+        return;
+      }
+
```

Comment:
You can automatically reconnect if no `queueName` have been passed.

Doing so mean I loose guarantees. If `queueName` is not given, `reconnect` should be an option that defaults to `false`.

Reply:
I think this library is basically not useful for exclusive queues.  I'll just remove support for that entirely.